axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/DefaultSegmentTransformer.java

```diff
@@ -126,55 +129,82 @@ private void closeTransactionIterator() {
-                        }
-                    } else if (!transactionIteratorRef.get().hasNext()) {
-                        done = true;
+                    Event event = originalTX.getEvent(transformedTransaction.size());
+                    if (transformedEvent.getToken() == originalTX.getToken() + transformedTransaction.size()) {
+                        event = transform(event, transformedEvent.getEvent());
+                        found = true;
                     }
                     transformedTransaction.add(event);
 
-                    if (originalTX.getEvents().size() == transformedTransaction.size()) {
-                        SegmentWriter segmentWriter = segmentWriterRef.get();
-                        segmentWriter.write(transformedTransaction);
-                        originalTransactionRef.set(null);
+                    if (transformedTransaction.size() == originalTX.getEventsCount()) {
+                        write(transformedTransaction);
                         transformedTransaction.clear();
+                        originalTransactionRef.set(null);
                     }
-                } while (!done);
+                } while (!found);
                 sink.success();
             } catch (Exception e) {
                 sink.error(e);
             }
         }).subscribeOn(Schedulers.boundedElastic());
     }
+
+    private Event transform(Event originalEvent, Event replacementEvent) {
+        if (Event.getDefaultInstance().equals(replacementEvent)) {
+            return Event.newBuilder()
+                        .setTimestamp(originalEvent.getTimestamp())
+                        .setAggregateIdentifier(originalEvent.getAggregateIdentifier())
+                        .setAggregateSequenceNumber(originalEvent.getAggregateSequenceNumber())
+                        .setAggregateType(originalEvent.getAggregateType().isEmpty() ? "" : "-")
+                        .setPayload(EMPTY_PAYLOAD)
+                        .build();
+        }
+        return replacementEvent;
+    }
+
+    private SerializedTransactionWithToken originalTransaction() {
+        if (originalTransactionRef.get() == null && transactionIteratorRef.get().hasNext()) {
+            originalTransactionRef.set(transactionIteratorRef.get().next());
+        }
+
+        return originalTransactionRef.get();
+    }
+
+    private void write(List<Event> events) throws IOException {
+        SegmentWriter segmentWriter = segmentWriterRef.get();
+        segmentWriter.write(events);
+    }
+
+    private void writeRemainingEvents() throws IOException {
+        // first complete the remaining events in the original transaction
+        SerializedTransactionWithToken originalTX = originalTransactionRef.get();
+        if (originalTX != null) {
+            while (transformedTransaction.size() < originalTX.getEventsCount()) {
+                transformedTransaction.add(originalTX.getEvent(transformedTransaction.size()));
```

Comment:
This will always add the last event?

Reply:
Changed the loop so the logic is easier to read.